beacon-chain/sync/pending_blocks_queue.go

```diff
@@ -0,0 +1,104 @@
+package sync
+
+import (
+	"context"
+	"encoding/hex"
+	"sort"
+	"time"
+
+	peer "github.com/libp2p/go-libp2p-peer"
+	"github.com/prysmaticlabs/go-ssz"
+	"github.com/prysmaticlabs/prysm/shared/bytesutil"
+	"github.com/prysmaticlabs/prysm/shared/params"
+	"github.com/sirupsen/logrus"
+)
+
+var processPendingBlocksPeriod = time.Duration(params.BeaconConfig().SecondsPerSlot/3) * time.Second
+
+// processes pending blocks queue on every processPendingBlocksPeriod
+func (r *RegularSync) processPendingBlocksQueue() {
+	ticker := time.NewTicker(processPendingBlocksPeriod)
+	for {
+		ctx := context.TODO()
+		select {
+		case <-ticker.C:
+			r.processPendingBlocks(ctx)
+		case <-r.ctx.Done():
+			log.Debug("p2p context is closed, exiting routine")
+			break
+
+		}
+	}
+}
+
+// processes the block tree inside the queue
+func (r *RegularSync) processPendingBlocks(ctx context.Context) error {
+	pids := r.peerIDs()
+	slots := r.sortedPendingSlots()
+
+	for _, s := range slots {
+		r.pendingQueueLock.RLock()
+		b := r.slotToPendingBlocks[uint64(s)]
+		inPendingQueue := r.seenPendingBlocks[bytesutil.ToBytes32(b.ParentRoot)]
+		r.pendingQueueLock.RUnlock()
+
+		inDB := r.db.HasBlock(ctx, bytesutil.ToBytes32(b.ParentRoot))
+		hasPeer := len(pids) != 0
+
+		// Only request for missing parent block if it's not in DB, not in pending cache
+		// and has peer in the peer list.
+		if !inPendingQueue && !inDB && hasPeer {
+			log.WithFields(logrus.Fields{
+				"currentSlot": b.Slot,
+				"parentRoot":  hex.EncodeToString(b.ParentRoot),
+			}).Info("Requesting parent block")
+			req := [][32]byte{bytesutil.ToBytes32(b.ParentRoot)}
+			if err := r.sendRecentBeaconBlocksRequest(ctx, req, pids[0]); err != nil {
```

Comment:
Do we only ask the first peer in the peerstore? Is that a good idea?

Reply:
oops, that's stubbed until we have round robin sync. Do I have the correct understanding the round robin sync implementation will export an API to return rotating peers? Can I add a todo for this until then?

Reply:
Yep, TODO sounds good. I don;t have a specific API in mind, but round robin sync is my next task so I'll keep this idea in mind.